### PR TITLE
[MRG+1] Add function for random parcellation

### PR DESCRIPTION
**Pull request by makkostya**
@agramfort @massich @mathurinm 
This PR add the new function to produce random parcellation of the cortex. It grows a number (input parameter) of random not intersecting cortical regions which cover the whole surface. 

Here is example of use (you should change annotation file names):

```
from pyface.qt import QtGui, QtCore
%matplotlib qt

import copy
import mne
from mne.datasets import sample
from surfer import Brain

subject_dir = sample.data_path() + '/subjects'
subject = 'sample'

n_parcel = 150
hemis = [0, 1]

# Parcellation
labels = mne.random_parcellation(subject, n_parcel, hemis, subject_dir, surface='white')

# Labels to annotations
annot_fname_lh = '/user/kmaksyme/home/tmp/lh.test.annot'
mne.label.write_labels_to_annot(labels,annot_fname=annot_fname_lh,hemi='lh',overwrite=True)
annot_fname_rh = '/user/kmaksyme/home/tmp/rh.test.annot'
mne.label.write_labels_to_annot(labels,annot_fname=annot_fname_rh,hemi='rh',overwrite=True)

# Visu
brain = Brain(subjects_dir=subject_dir,subject_id = subject,hemi="both",surf='inflated')
brain.add_annotation(annot_fname_lh, hemi='lh', borders=True)
brain.add_annotation(annot_fname_rh, hemi='rh', borders=True, remove_existing=False)
```

### CODE REVIEW

File: mne/label.py [hunk 1]
```diff
@@ -1659,6 +1660,159 @@ def _grow_nonoverlapping_labels(subject, seeds_, extents_, hemis, vertices_,
...
+    cover the whole surface. Regions are growing around randomly chosen
+    seeds.
+
+    Parameters
+    ----------
+    subject : string
+        Name of the subject as in SUBJECTS_DIR.
+    n_parcel : int
+        Total number of cortical parcels
+    hemis : array | int
```
Comment (agramfort) [inline mne/label.py]: it should be called hemi and be a string. Following the API of pysurfer:

https://github.com/nipy/PySurfer/blob/master/surfer/viz.py#L301

File: mne/label.py [hunk 2]
```diff
@@ -1659,6 +1660,159 @@ def _grow_nonoverlapping_labels(subject, seeds_, extents_, hemis, vertices_,
...
+        label.color = color
+
+    return labels
+
+
+def _cortex_parcellation(subject, n_parcel, hemis, vertices_, graphs):
+    """Random cortex parcellation."""
+    labels = []
+    parcel_size = np.floor((len(vertices_[hemis[0]]) +
+                            len(vertices_[hemis[1]])) / n_parcel)
```
Comment (agramfort) [inline mne/label.py]: I think

`len(vertices_[hemis[0]]) + len(vertices_[hemis[1]])) // n_parcel`

is nicer and is the same

File: mne/label.py [hunk 3]
```diff
@@ -1659,6 +1660,159 @@ def _grow_nonoverlapping_labels(subject, seeds_, extents_, hemis, vertices_,
...
+    parc_both = []
+    for hemi in set(hemis):
+        graph = graphs[hemi]  # distance graph
+        n_vertices = len(vertices_[hemi])
+
+        # prepare parcellation
+        parc = np.empty(n_vertices, dtype='int32')
+        parc[:] = -1
```
Comment (agramfort) [inline mne/label.py]: use np.full

https://docs.scipy.org/doc/numpy/reference/generated/numpy.full.html

File: mne/label.py [hunk 4]
```diff
@@ -1659,6 +1660,159 @@ def _grow_nonoverlapping_labels(subject, seeds_, extents_, hemis, vertices_,
...
+
+        # initialize active sources
+        s = 0
```
Comment (agramfort) [inline mne/label.py]: line is useless

File: mne/label.py [hunk 5]
```diff
@@ -1659,6 +1660,159 @@ def _grow_nonoverlapping_labels(subject, seeds_, extents_, hemis, vertices_,
...
+            parc[vertices] = label_id[j]
+            label_id = np.delete(label_id, i, 0)
+
+        # convert parc to labels
+        for i in xrange(n_labels):
+            vertices = np.nonzero(parc == label_id[i])[0]
+            name = 'label_' + str(i)
+            label_ = Label(vertices, hemi=hemi, name=name, subject=subject)
+            labels.append(label_)
+        parc_both += [parc]
```
Comment (agramfort) [inline mne/label.py]: is this the same algo as in brainstorm?

File: mne/label.py [hunk 6]
```diff
@@ -1659,6 +1660,159 @@ def _grow_nonoverlapping_labels(subject, seeds_, extents_, hemis, vertices_,
     return labels
 
 
+def random_parcellation(subject, n_parcel, hemis, subjects_dir=None,
+                        surface='white'):
```
Comment (agramfort) [inline mne/label.py]: you should use np.random and have a random_state parameter to make the result reproducible.

git grep random_state

to see how we do it.

**Comment by agramfort** [comment]
for the lazy curious is the output of the script shared by @makkostya 

screen shot 2018-02-23 at 18 45 14

**Comment by larsoner** [comment]
Oh the shame of having to run the example yourself to produce this output. It should be done by CircleCI and sphinx-gallery :)

File: mne/label.py (hunk 5 above)
Comment (makkostya) [reply mne/label.py]: I followed the same logic as brainsotm code:
 - define number of vertices per parcel
 - grow parcels till there are no free neighbor vertices or till riches max size
 - merge smallest parcels till rich the needed number

File: mne/label.py (hunk 1 above)
Comment (makkostya) [reply mne/label.py]: I used the doc of the function "grow_labels". If need to change it, probably should do it for grow_labels too.

**Comment by agramfort** [comment]
> I used the doc of the function "grow_labels". If need to change it,
probably should do it for grow_labels too.

hum indeed. I would recommend to use the surfer way and we'll deprecate the
hemis param in grow_label in another PR.

Having to pass array of int is really not pythonic

**Comment by makkostya** [comment]
Ok, I'll change it. And can do PR for "grow_labels" too

**Comment by agramfort** [comment]
thanks a lot

**Comment by makkostya** [comment]
> hum indeed. I would recommend to use the surfer way and we'll deprecate the hemis param in grow_label in another PR.

grow_label function has parameter hemis which is different from pysurf because we can associate the hemisphere to each source seed. So may be it makes sense to keep it like it is?
But I changed it in random_parcellation.

File: mne/label.py [hunk 7]
```diff
@@ -1659,6 +1660,162 @@ def _grow_nonoverlapping_labels(subject, seeds_, extents_, hemis, vertices_,
...
+    for hemi in set(hemis):
+        parcel_size = len(hemis) * len(vertices_[hemi]) // n_parcel
+        graph = graphs[hemi]  # distance graph
+        n_vertices = len(vertices_[hemi])
+
+        # prepare parcellation
+        parc = np.full(n_vertices, -1, dtype='int32')
+
+        # initialize active sources
+        rng = check_random_state(random_state)
```
Comment (agramfort) [inline mne/label.py]: I would have put the check_random_state outside of the for loop on hemi no?

Comment (makkostya) [reply mne/label.py]: Yes, done.

**Comment by agramfort** [comment]
LGTM 

any other review?

**Comment by larsoner** [comment]
LGTM but @makkostya your `git` / GitHub config appears to be incorrect. I should be able to click on the little icon to the left of your commits to get to your GitHub profile. It might be a matter of adding your `git config` email address to your GitHub account, not sure. Do you want to fix it before merge?

**Comment by makkostya** [comment]
@larsoner , I added my `git config` email address to my GitHub account, is it Ok now?

**Comment by larsoner** [comment]
Yep all better, thanks @makkostya

**Pull request merged**